src/image/ExposureFitsReader.cc

```diff
@@ -48,8 +52,155 @@ bool _contains(std::array<T, N> const& array, T const& value) {
+    // FilterLabel::from* returns a statically allocated object, so only way
+    // to get it into shared_ptr is to copy it.
+    if (_isBand(name)) {
+        return std::make_shared<FilterLabel>(FilterLabel::fromBand(name));
+    } else {
+        return std::make_shared<FilterLabel>(FilterLabel::fromPhysical(name));
+    }
+}
+
+/**
+ * Convert an old-style Filter to a FilterLabel.
+ *
+ * @param filter Any Filter object.
+ * @returns The closest equivalent FilterLabel, given available information.
+ */
+// TODO: compatibility code to be removed in DM-27177
+std::shared_ptr<FilterLabel> makeFilterLabel(Filter const& filter) {
+    // Filter has no self-consistency guarantees whatsoever, and most methods
+    // are unsafe. Program extremely defensively.
+    if (filter.getId() == Filter::UNKNOWN) {
+        // Not a registered filter; guarantees on canonical name do not apply
+        return makeFilterLabelDirect(filter.getName());
+    }
+
+    // obs.base.FilterDefinition ensures the canonical name is the first defined
+    // of afwname, band name, physical filter name.
+    std::string canonical;
+    try {
+        canonical = filter.getCanonicalName();
+    } catch (pex::exceptions::NotFoundError const&) {
+        // Not a registered filter; guarantees on canonical name do not apply
+        return makeFilterLabelDirect(filter.getName());
+    }
+
+    if (_AFW_NAMES.count(canonical) > 0) {
+        return std::make_shared<FilterLabel>(_AFW_NAMES.at(canonical));
+    } else if (_isBand(canonical)) {
+        // physical filter is one of the aliases, but can't tell which one
+        // (FilterDefinition helpfully sorts them). Safer to leave it blank.
+        std::vector<std::string> aliases;
+        try {
+            aliases = filter.getAliases();
+        } catch (pex::exceptions::NotFoundError const&) {
+            // No aliases; leave the vector empty
+        }
+        if (aliases.size() == 1) {
+            return std::make_shared<FilterLabel>(FilterLabel::fromBandPhysical(canonical, aliases.front()));
+        } else {
+            return std::make_shared<FilterLabel>(FilterLabel::fromBand(canonical));
+        }
+    } else {
+        return std::make_shared<FilterLabel>(FilterLabel::fromPhysical(canonical));
+    }
+}
+
+/**
+ * Convert an old-style single Filter name to a FilterLabel, using available information.
+ *
+ * @param name The name persisted in a FITS file. May be any of a Filter's many names.
+ * @returns The closest equivalent FilterLabel, given available information.
+ */
+std::shared_ptr<FilterLabel> makeFilterLabel(std::string const& name) {
+    if (_AFW_NAMES.count(name) > 0) {
+        return std::make_shared<FilterLabel>(_AFW_NAMES.at(name));
+    }
+    // else name is either a band, a physical filter, or a deprecated alias
+
+    try {
+        // To ease the transition to FilterLabel, use Filter to get all the names
+        // TODO: after DM-27177, leave only the catch block
```

Comment:
Maybe worth noting here (or in `makeFilterLabel(Filter)` above?) that what makes this work is the pre-registration of filters by `FilterDefinition`, so that there is extra information available from the singleton. It took me a bit to sort out how you were getting that extra information, and it would be very non-obvious to someone who isn't familiar with `Filter`.